lib/full-site-editing/block-templates.php

```diff
@@ -0,0 +1,333 @@
+ *
+ * @return array Template.
+ */
+function _gutenberg_get_template_files( $template_type ) {
+	$template_base_paths = array(
+		'wp_template'      => 'block-templates',
+		'wp_template_part' => 'block-template-parts',
+	);
+	$themes              = array(
+		get_stylesheet() => get_stylesheet_directory(),
+		get_template()   => get_template_directory(),
+	);
+
+	$template_files = array();
+	foreach ( $themes as $theme_slug => $theme_dir ) {
+		$theme_template_files = _gutenberg_get_template_paths( $theme_dir . '/' . $template_base_paths[ $template_type ] );
+		foreach ( $theme_template_files as $template_file ) {
+			$template_base_path = $template_base_paths[ $template_type ];
+			$template_slug      = substr(
+				$template_file,
+				// Starting position of slug.
+				strpos( $template_file, $template_base_path . DIRECTORY_SEPARATOR ) + 1 + strlen( $template_base_path ),
+				// Subtract ending '.html'.
+				-5
+			);
+			$template_files[] = array(
+				'slug'  => $template_slug,
+				'path'  => $template_file,
+				'theme' => $theme_slug,
+				'type'  => $template_type,
+			);
+		}
+	}
+
+	return $template_files;
+}
+
+/**
+ * Build a unified template object based on a theme file.
+ *
+ * @param array $template_file Theme file.
+ * @param array $template_type wp_template or wp_template_part.
+ *
+ * @return WP_Block_Template Template.
+ */
+function _gutenberg_build_template_result_from_file( $template_file, $template_type ) {
+	$default_template_types = gutenberg_get_default_template_types();
+
+	$theme               = wp_get_theme()->get_stylesheet();
+	$template            = new WP_Block_Template();
+	$template->id        = $theme . '//' . $template_file['slug'];
+	$template->theme     = $theme;
+	$template->content   = file_get_contents( $template_file['path'] );
+	$template->slug      = $template_file['slug'];
+	$template->is_custom = false;
+	$template->type      = $template_type;
+	$template->title     = $template_file['slug'];
+	$template->status    = 'publish';
+
+	if ( 'wp_template' === $template_type && isset( $default_template_types[ $template_file['slug'] ] ) ) {
+		$template->description = $default_template_types[ $template_file['slug'] ]['description'];
+		$template->title       = $default_template_types[ $template_file['slug'] ]['title'];
+	}
+
+	return $template;
+}
+
+/**
+ * Build a unified template object based a post Object.
+ *
+ * @param WP_Post $post Template post.
+ *
+ * @return WP_Block_Template|WP_Error Template.
+ */
+function _gutenberg_build_template_result_from_post( $post ) {
+	$terms = get_the_terms( $post, 'wp_theme' );
+
+	if ( is_wp_error( $terms ) ) {
+		return $terms;
+	}
+
+	if ( ! $terms ) {
+		return new WP_Error( 'template_missing_theme', __( 'No theme is defined for this template.', 'gutenberg' ) );
+	}
+
+	$theme = $terms[0]->name;
+
+	$template              = new WP_Block_Template();
+	$template->wp_id       = $post->ID;
+	$template->id          = $theme . '|' . $post->post_name;
```

Comment:
I think this is the only one missing in the actual code!

```suggestion
	$template->id          = $theme . '//' . $post->post_name;
```

It also occurs in `class-block-templates-test.php` a few times.

Reply:
(Also apologies for the multiple separate change requests! I accidentally submitted the review _twice_... I'll try to be more carefule now.)